lib/global-styles.php

```diff
@@ -147,155 +147,325 @@ function gutenberg_experimental_global_styles_get_user_cpt_id() {
+		$theme_presets,
+		$theme_config
+	);
+
+	return $theme_config;
+}
+
+/**
+ * Retrieves the block data (selector/supports).
+ *
+ * @return array
+ */
+function gutenberg_experimental_global_styles_get_block_data() {
+	// TODO: this data should be taken from the block registry.
+	//
+	// At the moment this array replicates the current capabilities
+	// declared by blocks via __experimentalLineHeight,
+	// __experimentalColor, and __experimentalFontSize.
+	$block_data = array(
+		'global'          => array(
+			'selector' => ':root',
+			'supports' => array(),
+		),
+		'core/paragraph'  => array(
+			'selector' => 'p',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h1' => array(
+			'selector' => 'h1',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h2' => array(
+			'selector' => 'h2',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h3' => array(
+			'selector' => 'h3',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h4' => array(
+			'selector' => 'h4',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h5' => array(
+			'selector' => 'h5',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/heading/h6' => array(
+			'selector' => 'h6',
+			'supports' => array( 'line-height', 'font-size', 'color' ),
+		),
+		'core/columns'    => array(
+			'selector' => '.wp-block-columns',
+			'supports' => array( 'color' ),
+		),
+		'core/group'      => array(
+			'selector' => '.wp-block-group',
+			'supports' => array( 'color' ),
+		),
+		'core/media-text' => array(
+			'selector' => '.wp-block-media-text',
+			'supports' => array( 'color' ),
 		),
-		$theme_supports
 	);
+
+	return $block_data;
+}
+
+/**
+ * Takes a global styles tree and generates
+ * the corresponding stylesheet.
+ *
+ * @param array $tree Global Styles tree.
+ *
+ * @return string
+ */
+function gutenberg_experimental_global_styles_resolver( $tree ) {
+	$stylesheet = '';
+	$block_data = gutenberg_experimental_global_styles_get_block_data();
+	foreach ( array_keys( $tree ) as $block_name ) {
+		if (
+			! array_key_exists( $block_name, $block_data ) ||
+			! array_key_exists( 'selector', $block_data[ $block_name ] ) ||
+			! array_key_exists( 'supports', $block_data[ $block_name ] )
+		) {
+			// Skip blocks that haven't declared support,
+			// so we don't know to process.
+			continue;
+		}
+
+		$token         = '--';
+		$prefix        = '--wp--preset' . $token;
+		$css_variables = gutenberg_experimental_global_styles_get_css_vars( $tree[ $block_name ]['presets'], $prefix, $token );
+
+		$stylesheet .= gutenberg_experimental_global_styles_resolver_styles(
+			$block_data[ $block_name ]['selector'],
+			$block_data[ $block_name ]['supports'],
+			array_merge( $tree[ $block_name ]['styles'], $css_variables )
```

Comment:
This means that if a block declares presets, they'll be attached to the CSS rule as any other property.

Input: 

```
{
    "core/group": {
        "styles": {
            "color": "var(--wp--preset--color--primary)"
        },
        "presets": {
            "color": [
                 "primary": "#abb8c3"
            ]
        }
    }
}
```

Output:

```css
.wp-block-group { /* selector taken from block registry */
    color: var(--wp--preset--color--primary);
    --wp--preset--color--primary: #abb8c3;
}
```